src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/PackageRestore/PackageRestoreDataSource.cs

```diff
@@ -185,6 +185,22 @@ private RestoreData CreateRestoreData(ProjectRestoreInfo restoreInfo, bool succe
                 succeeded: succeeded && lastWriteTime != DateTime.MinValue);
         }
 
+        protected virtual bool IsRestoreDataVersionOutOfDate(IImmutableDictionary<NamedIdentity, IComparable> dataVersions)
+        {
+            Assumes.Present(_project.Services.DataSourceRegistry);
+
+            IProjectDataSourceRegistry dataSourceRegistry = _project.Services.DataSourceRegistry;
+            foreach (KeyValuePair<NamedIdentity, IComparable> versionDescription in dataVersions)
+            {
+                if (dataSourceRegistry.TryGetDataSource(versionDescription.Key, out IProjectValueDataSource? dataSource) && versionDescription.Value.CompareTo(dataSource.DataSourceVersion) < 0)
+                {
+                    return true;
+                }
+            }
```

Comment:
I thought CPS had utility methods for comparing versions against a registry. But if it does, you probably wrote them and would have known about them :)

Reply:
Actually I think I'm remembering the implementation of `GetLatestVersionAsync`.

Reply:
that is fair point, we should provide something suitable here, more general than GetLatestVersionAsync.

I think this logic is still not enough to ensure all target platforms are loaded. I will take a look and discuss with you what we can do safely.  when Richard starts to enable async path to load solution, we now have more chances to run into new race conditions.